teachertool/src/services/makecodeEditorService.ts

```diff
@@ -0,0 +1,52 @@
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+let makecodeEditorRef: HTMLIFrameElement | undefined;
+const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+let nextId: number = 0;
+let pendingMessages: {[index: string]: PendingMessage} = {};
+
+const onMessageReceived = (event: MessageEvent) => {
+    const data = event.data as pxt.editor.EditorMessageRequest;
+    if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+        const pending = pendingMessages[data.id];
+        pending.handler(data);
+        delete pendingMessages[data.id];
+        return;
+    }
+}
+
+export const sendMessageAsync = (message?: any) => {
+    return new Promise(resolve => {
+        const sendMessageCore = (message: any) => {
+            message.response = true;
+            message.id = nextId++ + "";
+            pendingMessages[message.id] = {
+                original: message,
+                handler: resolve
+            };
+            makecodeEditorRef!.contentWindow!.postMessage(message, "*");
+        }
+
+        if (makecodeEditorRef) {
+            messageQueue.push(message);
```

Comment:
If I understand the goal correctly, we want the messageQueue to be used to store messages that came in before the iframe was ready. With that in mind, I think we want to push to the queue when the iframe _is not_ ready, then clear the queue when it is. I think the code as written would just send the message twice when the iframe is ready and then lose any messages that came in while the iframe was not ready.

Reply:
I think we'll need to do one of:
A. queue the messages until the iframe is ready to receive them.
B. during app init, keep the UI in a loading state until the iframe is ready.

Reply:
I think we'd want to do both A and B, ideally

Reply:
Though B could be done in a separate change, IMO

Reply:
Makes sense. I just did some testing with this. The frame gets ready pretty quickly, so there's not much wait time, but I put a five second timeout before attaching the ref to the service and was able to stack up a lot of requests. I just wanted to double check that what the desired clear mechanism is. Because with the playing that I did within `sendMessageAsync`, all of the message wouldn't be processed until another message was sent to the iframe. I think this would be the job of an event listener that we were chatting about before. So we have the window listen for when the frame is ready, and when it is, send the messages over.

Reply:
Kinda related to this work stream, I don't see the value in `if (message) sendMessageCore(message)` it seems like it duplicates calls..  We push the latest message to the queue and then clear the queue, so shouldn't the most recent message already have called `sendMessageCore()`?

Reply:
That's the reason behind my note that this would just send the message twice, which is not what we want. I think you could keep `if (message) sendMessageCore(message)` but remove the part where it adds it to the queue first. So you clear the queue (without adding this message), then send this message.

As for your note about "all of the message wouldn't be processed until another message was sent to the iframe", I believe that's the reasoning behind this comment: https://github.com/microsoft/pxt/pull/9811#discussion_r1442057193 Basically, we want to flush the queue the moment we learn that the iframe is ready, in addition to anytime a new message is sent.

Reply:
Right, makes sense. 

I think I would prefer to keep the pushing to the queue logic and get rid of that `if (message)...` stuff because of this comment of yours:

>With that in mind, I think we want to push to the queue when the iframe is _not_ ready, then clear the queue when it is.

Because I can move `messageQueue.push(message)` outside of the `if (makecodeEditorRef)` check and allow for pushing messages whether or not the iframe is ready. This way, the logic stays consistent. When the iframe is ready, we can flush the queue with another function/event listener and when the iframe is ready. Separately, when a message is sent, the queue will just get cleared immediately. 

I guess the possible problem here would be a race condition, though, in the queue.

Reply:
Spoke offline, but to reiterate here: I don't think we really need to worry about race conditions in this scenario. Your proposed approach seems fine, as long as we still make sure to send all pending messages once the iframe is ready and before any new messages get sent.

Reply:
Oh but small gotcha: make sure you only add the message to the queue if it's not undefined. (In skillmap, at least, calling sendMessageAsync with undefined clears the queue without sending a new message. If we're following that pattern, that case is important to account for)